File: 2023/08.07/정호조/1학년.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 일학년_5557 {
    static long[][] dp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        dp = new long[N][21];
        int[] num = new int[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            num[i] = Integer.parseInt(st.nextToken());
        }

        dp[0][num[0]] = 1;

        for (int i = 0; i < N - 1; i++) {
            for (int j = 0; j <= 20; j++) {
                if (dp[i][j] != 0) {
                    int dif = j - num[i + 1];
                    int sum = j + num[i + 1];

                    if (0 <= dif && dif <= 20) dp[i + 1][dif] += dp[i][j];
                    if (0 <= sum && sum <= 20) dp[i + 1][sum] += dp[i][j];
                }
            }
        }

//        for (int i = 0; i < N; i++) {
//            System.out.println(Arrays.toString(Arrays.stream(dp[i]).toArray()));
//        }

        System.out.println(dp[N - 2][num[N - 1]]);
    }
}

File: 2023/08.07/조수연/1학년.kt
import java.io.*
import java.util.*

lateinit var st: StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = readLine().toInt()
    val dp = Array(n) { LongArray(21) }
    val numArr = IntArray(n)

    st = StringTokenizer(readLine())

    repeat(n) {
        numArr[it] = st.nextToken().toInt()
    }

    dp[0][numArr[0]] = 1 // 초깃값

    var plus = 0
    var minus = 0

    for (i in 1 until n - 1) { // 등호(=)를 제외하고 덧셈과 뺄셈의 횟수는 n-2번
        for (j in 0..20) {
            if (dp[i - 1][j] != 0L) { // 계산하기 전의 값
                plus = j + numArr[i]
                minus = j - numArr[i]

                if (plus in 0..20) {
                    dp[i][plus] += dp[i - 1][j]
                }

                if (minus in 0..20) {
                    dp[i][minus] += dp[i - 1][j]
                }
            }
        }
    }

    bw.write("${dp[n - 2][numArr[n - 1]]}")

    bw.flush()
    bw.close()
}